app/scripts/modules/core/src/pipeline/config/stages/entityTags/TagEditor.tsx

```diff
@@ -0,0 +1,169 @@
+  onChange: (tag: IEntityTag) => void;
+  tag: IEntityTag;
+}
+
+export interface ITagEditorState {
+  type: EntityTagType;
+  valueIsObject: boolean;
+}
+
+const typeOptions: Array<Option<EntityTagType>> = [
+  { label: 'Notice', value: 'notice' },
+  { label: 'Alert', value: 'alert' },
+  { label: 'Custom', value: 'custom' },
+];
+
+export class TagEditor extends React.Component<ITagEditorProps, ITagEditorState> {
+  constructor(props: ITagEditorProps) {
+    super(props);
+
+    const type = props.tag.name.startsWith('spinnaker_ui_notice')
+      ? 'notice'
+      : props.tag.name.startsWith('spinnaker_ui_alert')
+        ? 'alert'
+        : 'custom';
+
+    const valueIsObject = typeof props.tag.value === 'object';
+
+    this.state = {
+      type,
+      valueIsObject,
+    };
+  }
+
+  public componentWillReceiveProps(nextProps: ITagEditorProps) {
+    const valueIsObject = typeof nextProps.tag.value === 'object';
+    this.setState({ valueIsObject });
+  }
+
+  private handleTypeChanged = (type: EntityTagType) => {
+    this.setState({ type });
+
+    const tag = { ...this.props.tag };
+    if (type !== 'custom') {
+      delete tag.namespace;
+      tag.name = `spinnaker_ui_${type}:${UUIDGenerator.generateUuid()}`;
+      tag.value = {
+        type,
+        message: '',
+      };
+    } else {
+      tag.name = '';
+      tag.value = '';
+    }
+    this.props.onChange(tag);
+  };
+
+  private tagValueChanged = (key: string, value: any) => {
+    if (key === 'value') {
+      try {
+        value = JSON.parse(value);
+      } catch (e) {}
+    }
+    const tag = { ...this.props.tag };
+    set(tag, key, value);
+    this.props.onChange(tag);
+  };
+
+  public render() {
+    const { tag } = this.props;
+    const { type, valueIsObject } = this.state;
+
+    const value = valueIsObject ? JSON.stringify(this.props.tag.value) : this.props.tag.value;
+
+    const namespaceInput =
+      type === 'custom' ? (
+        <div className="row">
+          <label className="col-md-3 sm-label-right">
+            Namespace <HelpField id="pipeline.config.entitytags.namespace" />
+          </label>
+          <div className="col-md-8">
+            <input
+              className="form-control input-sm"
+              value={tag.namespace || ''}
+              onChange={event => this.tagValueChanged('namespace', event.target.value)}
+              type="text"
+            />
+          </div>
+        </div>
+      ) : null;
+
+    const nameInput =
+      type === 'custom' ? (
+        <div className="row">
+          <label className="col-md-3 sm-label-right">Name</label>
+          <div className="col-md-8">
+            <input
+              className="form-control input-sm"
+              value={tag.name || ''}
+              onChange={event => this.tagValueChanged('name', event.target.value)}
+              type="text"
+            />
+          </div>
+        </div>
+      ) : null;
+
+    const valueInput =
```

Comment:
I find these ternaries a bit hard to read.  Maybe `renderFoo()` methods would read better?

Reply:
Happy to change it, but not sure what you find hard to read or what exactly you mean by a `renderFoo()` method...

Reply:
I guess I meant something like this:

```
private renderValueInput(value: any)
      if (type === 'custom') {
        return (
          <div className="row">
          <label className="col-md-3 sm-label-right">
            Value <HelpField id="pipeline.config.entitytags.value" />
          </label>
          <div className="col-md-8">
            <FormField
              input={TextInput}
              value={value || ''}
              onChange={event => this.tagValueChanged('value', event.target.value)}
            />
          </div>
        </div>
      ) 

      return (
        <div className="row">
          <label className="col-md-3 sm-label-right">Message</label>
          <div className="col-md-8">
            <FormField
              input={TextInput}
              value={tag.value.message || ''}
              onChange={event => this.tagValueChanged('value.message', event.target.value)}
            />
          </div>
        </div>
      );
}
```

but tbh I'm not sure how much better it is.  I think the ternaries are slightly harder to scan

Reply:
I've found that most ternaries where the JSX blocks are > ~4 lines get tough to scan if you don't already know the structure. Also makes it easier to throw more logic in later without needing to contort yourself into a bunch of extra ternaries. That being said it's a minor annoyance, don't twist yourself in knots.

Reply:
since no strong opinions, leaving as is for now.